quickwit/quickwit-indexing/src/source/queue_sources/message.rs

```diff
@@ -0,0 +1,295 @@
+        message_type: MessageType,
+    ) -> Result<PreProcessedMessage, PreProcessingError> {
+        let payload = match message_type {
+            MessageType::S3Notification => PreProcessedPayload::ObjectUri(
+                uri_from_s3_notification(&self.payload, &self.metadata.ack_id)?,
+            ),
+            MessageType::RawUri => {
+                let payload_str = read_to_string(self.payload).context("failed to read payload")?;
+                PreProcessedPayload::ObjectUri(Uri::from_str(&payload_str)?)
+            }
+        };
+        Ok(PreProcessedMessage {
+            metadata: self.metadata,
+            payload,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum PreProcessedPayload {
+    ObjectUri(Uri),
+}
+
+impl PreProcessedPayload {
+    pub fn partition_id(&self) -> PartitionId {
+        match &self {
+            Self::ObjectUri(uri) => PartitionId::from(uri.as_str()),
+        }
+    }
+}
+
+/// A message that went through the minimal transformation to discover its
+/// partition id. Indeed, the message might be discarded if the partition was
+/// already processed, so it's better to avoid doing unnecessary work at this
+/// stage.
+#[derive(Debug, PartialEq, Eq)]
+pub struct PreProcessedMessage {
+    pub metadata: MessageMetadata,
+    pub payload: PreProcessedPayload,
+}
+
+impl PreProcessedMessage {
+    pub fn partition_id(&self) -> PartitionId {
+        self.payload.partition_id()
+    }
+}
+
+fn uri_from_s3_notification(message: &OwnedBytes, ack_id: &str) -> Result<Uri, PreProcessingError> {
+    let value: Value =
+        serde_json::from_slice(message.as_slice()).context("invalid JSON message")?;
+    if matches!(value["Event"].as_str(), Some("s3:TestEvent")) {
+        return Err(PreProcessingError::Discardable {
+            reason: "S3 test event",
+            ack_id: ack_id.to_string(),
+        });
+    }
+    let key = value["Records"][0]["s3"]["object"]["key"]
+        .as_str()
+        .context("invalid S3 notification: Records[0].s3.object.key not found")?;
+    let bucket = value["Records"][0]["s3"]["bucket"]["name"]
+        .as_str()
+        .context("invalid S3 notification: Records[0].s3.bucket.name not found".to_string())?;
+    Uri::from_str(&format!("s3://{}/{}", bucket, key)).map_err(|e| e.into())
+}
+
+/// A message for which we know as much of the global processing status as
+/// possible and that is now ready to be processed.
+pub struct ReadyMessage {
+    pub position: Position,
+    pub content: PreProcessedMessage,
+    pub visibility_handle: VisibilityTaskHandle,
+}
+
+impl ReadyMessage {
+    pub async fn start_processing(
+        self,
+        storage_resolver: &StorageResolver,
+    ) -> anyhow::Result<Option<InProgressMessage>> {
+        let partition_id = self.partition_id();
+        match self.content.payload {
+            PreProcessedPayload::ObjectUri(uri) => {
+                let batch_reader = ObjectUriBatchReader::try_new(
```

Comment:
What is happening here if the object is not present (was deleted for instance).

I think we need to log a warning and acknowledge the object to remove it from the queue.

Feel free to discuss the specification, but we need something that makes sense + a unit test checking that?

Reply:
I think a deleted message **is** an error. It should really never happen, and if it does we are in a situation where we want to let the user determine manually what to do. The safe way to this is to let the message go into a dead letter queue after a few retries.